diff -r 912f387906b7 backends/FTPStorage.py
--- a/backends/FTPStorage.py	Mon Mar 02 01:05:37 2009 +0100
+++ b/backends/FTPStorage.py	Wed May 13 13:36:35 2009 +0200
@@ -132,12 +132,15 @@
         
     def _save(self, name, content):
         content.open()
-        if hasattr(content, 'chunks'):
-            content_str = ''.join(chunk for chunk in content.chunks())
-        else:
-            content_str = content.read()
         self._start_connection()
-        self._put_file(name, content_str)
+        try:
+            self._mkremdirs(os.path.dirname(name))
+            pwd = self._connection.pwd()
+            self._connection.cwd(os.path.dirname(name))
+            self._connection.storbinary('STOR ' + os.path.basename(name), content.file, content.DEFAULT_CHUNK_SIZE)
+            self._connection.cwd(pwd)
+        except ftplib.all_errors, e:
+            raise FTPStorageException('Error writing file %s' % name)
         return name
 
     def _get_dir_details(self, path):
@@ -181,12 +184,15 @@
     def exists(self, name):
         self._start_connection()
         try:
-            if name in self._connection.nlst(os.path.dirname(name)):
+            if os.path.basename(name) in self._connection.nlst(os.path.dirname(name) + '/'):
                 return True
             else:
                 return False
         except ftplib.error_temp, e:
             return False
+        except ftplib.error_perm, e:
+            # error_perm: 550 Can't find file
+            return False
         except ftplib.all_errors, e:
             raise FTPStorageException('Error when testing existence of %s' % name)            
 
